src/Mod/Sketcher/App/SketchObject.cpp

```diff
@@ -1019,68 +1020,128 @@ int SketchObject::movePoint(int GeoId, PointPos PosId, const Base::Vector3d& toP
-        const Part::GeomLineSegment *lineSeg = static_cast<const Part::GeomLineSegment*>(geo);
-        if (PosId == PointPos::start)
-            return lineSeg->getStartPoint();
-        else if (PosId == PointPos::end)
-            return lineSeg->getEndPoint();
-    } else if (geo->is<Part::GeomCircle>()) {
-        const Part::GeomCircle *circle = static_cast<const Part::GeomCircle*>(geo);
-        auto pt = circle->getCenter();
-        if(PosId != PointPos::mid)
-             pt.x += circle->getRadius();
-        return pt;
-    } else if (geo->is<Part::GeomEllipse>()) {
-        const Part::GeomEllipse *ellipse = static_cast<const Part::GeomEllipse*>(geo);
-        auto pt = ellipse->getCenter();
-        if(PosId != PointPos::mid)
-            pt += ellipse->getMajorAxisDir()*ellipse->getMajorRadius();
-        return pt;
-    } else if (geo->is<Part::GeomArcOfCircle>()) {
-        const Part::GeomArcOfCircle *aoc = static_cast<const Part::GeomArcOfCircle*>(geo);
-        if (PosId == PointPos::start)
-            return aoc->getStartPoint(/*emulateCCW=*/true);
-        else if (PosId == PointPos::end)
-            return aoc->getEndPoint(/*emulateCCW=*/true);
-        else if (PosId == PointPos::mid)
-            return aoc->getCenter();
-    } else if (geo->is<Part::GeomArcOfEllipse>()) {
-        const Part::GeomArcOfEllipse *aoc = static_cast<const Part::GeomArcOfEllipse*>(geo);
-        if (PosId == PointPos::start)
-            return aoc->getStartPoint(/*emulateCCW=*/true);
-        else if (PosId == PointPos::end)
-            return aoc->getEndPoint(/*emulateCCW=*/true);
-        else if (PosId == PointPos::mid)
-            return aoc->getCenter();
-    } else if (geo->is<Part::GeomArcOfHyperbola>()) {
-        const Part::GeomArcOfHyperbola *aoh = static_cast<const Part::GeomArcOfHyperbola*>(geo);
-        if (PosId == PointPos::start)
-            return aoh->getStartPoint();
-        else if (PosId == PointPos::end)
-            return aoh->getEndPoint();
-        else if (PosId == PointPos::mid)
-            return aoh->getCenter();
-    } else if (geo->is<Part::GeomArcOfParabola>()) {
-        const Part::GeomArcOfParabola *aop = static_cast<const Part::GeomArcOfParabola*>(geo);
-        if (PosId == PointPos::start)
-            return aop->getStartPoint();
-        else if (PosId == PointPos::end)
-            return aop->getEndPoint();
-        else if (PosId == PointPos::mid)
-            return aop->getCenter();
-    } else if (geo->is<Part::GeomBSplineCurve>()) {
-        const Part::GeomBSplineCurve *bsp = static_cast<const Part::GeomBSplineCurve*>(geo);
-        if (PosId == PointPos::start)
-            return bsp->getStartPoint();
-        else if (PosId == PointPos::end)
-            return bsp->getEndPoint();
+        return getPointForGeometry<Part::GeomPoint>(static_cast<const Part::GeomPoint*>(geo), PosId);
+    }
+    else if (geo->is<Part::GeomLineSegment>()) {

```

Comment:
Shouldn't we remove all the tests and use the template argument deduction?

Reply:
Is there a way to get the `Part::GeomLineSegment` part out? This kind of test is happening because 1) we are given a `Part::Geometry`; 2) it is not easily obtainable what subclass of `Part::Geometry` it is.

I do picture abstracting this check itself through templates so that it can be used in other places like `trim`, `split`, `extend`, `exposeInternalGeometry` etc.

Reply:
You're right, test is needed. But I think template use unnecessarily heavy and overload sufficient.
```cpp
Base::Vector3d SketchObject::getPointForGeometry(const Part::Geometry *geo, PointPos PosId){
    if (geo->is<Part::GeomPoint>()) {
        return getPointForGeometry(static_cast<const Part::GeomPoint*>(geo), PosId);
    }
    else if (geo->is<Part::GeomLineSegment>()) {
        return getPointForGeometry(static_cast<const Part::GeomLineSegment*>(geo), PosId);
    }
    // ...
}
```

Reply:
Indeed. This was pointed out to me in some private channels as well, and most likely I will change this behavior sooner than later.

Overall, this is my plan:
```cpp
Base::Vector3d SketchObject::getPoint(const Part::Geometry *geo, PointPos PosId)
{
    template <class geomType>
        auto singleArgFunction = [&GeoId, &point](T* geo) {
            return getPointForGeometry<geomType>(GeoId, point);
        };

    return performActionByGeomType<Base::Vector3d, singleArgFunction>(geo);
}

int SketchObject::trim(int GeoId, const Base::Vector3d& point)
{
    template <class geomType>
        auto singleArgFunction = [&GeoId, &point](T* geo) {
            return trimByGeomType<geomType>(GeoId, point);
        };

    return performActionByGeomType<int, singleArgFunction>(getGeometry(GeoId));
}

template <returnType, singleArgFunction>
returnType SketchObject::performActionByGeomType(Part::Geometry* geo)
{
    if (geo->is<Part::GeomPoint>()) {
        return singleArgFunction<Part::GeomPoint>(static_cast<Part::GeomPoint*>(geo));
    }
    else if (geo->is<Part::GeomLineSegment>()) {
        return singleArgFunction<Part::GeomLineSegment>(static_cast<const Part::GeomLineSegment*>(geo));
    }
    // ... other supported cases

    return singleArgFunction<Part::Geometry>(geo);
}
```